### PR TITLE
 Enhance Pause Mechanism Security

### PR DESCRIPTION
**Pull request by NehharShah**
## Key Changes
1. Reduced Pause Cooldown Period
- Before: 1 week cooldown between pause operations
- After: 48 hours cooldown

2. Enhanced Exploit Prevention
- Before: Tracked lastPaused timestamp (updated when pausing)
- After: Track lastUnpaused timestamp (only updated when unpausing)

3. Timelock Protection for Authority Changes
- New Feature: 48-hour mandatory delay for pauseAuthority changes

## New Functions
- proposePauseAuthorityChange(address newPauseAuthority) - Propose authority change with timelock
- executePauseAuthorityChange() - Execute change after timelock expires
- getProposedPauseAuthorityChange() - View proposed changes

##  Testing
-  Updated existing testPause() for new behavior
- Added comprehensive testPauseAuthorityTimelock() test
- All pause-related tests passing
- Verified timelock edge cases and authorization checks

### CODE REVIEW

File: src/PauseAuthority.sol [hunk 1]
```diff
@@ -56,19 +74,53 @@ abstract contract PauseAuthority {
...
+        _pauseConfig = (uint256(lastUnpaused) << 160) | uint160(proposedAuthority);
+        
+        // Clear the proposed change
+        _proposedPauseAuthorityChange = 0;
+
+        emit PauseAuthoritySet(proposedAuthority);
+    }
+
+    /// @dev Legacy function for backward compatibility - now initiates the timelock process.
+    function setPauseAuthority(address newPauseAuthority) public virtual {
```
Comment (legion2002) [inline src/PauseAuthority.sol:123]: We don't need to retain the legacy function. Can just delete it.

File: src/PauseAuthority.sol [hunk 2]
```diff
@@ -5,49 +5,67 @@ abstract contract PauseAuthority {
...
+    error TimelockNotExpired();
+
     /// @dev The pause flag has been updated.
     event PauseSet(bool indexed isPaused);
 
     /// @dev The pause authority has been set to `pauseAuthority`.
     event PauseAuthoritySet(address indexed pauseAuthority);
 
+    /// @dev A new pause authority change has been proposed.
+    event PauseAuthorityChangeProposed(address indexed newPauseAuthority, uint256 indexed effectiveTime);
```
Comment (legion2002) [inline src/PauseAuthority.sol:18]: Can we change this event to `PauseAuthorityProposed(address indexed newPauseAuthority)`.The effective time can be deduced offchain from the timestamp of the log.

File: src/PauseAuthority.sol [hunk 3]
```diff
@@ -5,49 +5,67 @@ abstract contract PauseAuthority {
...
+
     /// @dev Time period after which the contract can be unpaused by anyone.
     uint256 public constant PAUSE_TIMEOUT = 4 weeks;
 
+    /// @dev Cooldown period before the contract can be paused again (reduced from 1 week to 48 hours).
```
Comment (legion2002) [inline src/PauseAuthority.sol:23]: Feel free to remove comments about maintaining legacy information. Contracts are in beta, and we're comfortable making breaking changes.

File: src/PauseAuthority.sol [hunk 4]
```diff
@@ -56,19 +74,53 @@ abstract contract PauseAuthority {
...
 
-    function setPauseAuthority(address newPauseAuthority) public virtual {
-        (address authority, uint40 lastPaused) = getPauseConfig();
+    /// @dev Returns the proposed pause authority change details.
+    function getProposedPauseAuthorityChange() public view virtual returns (address, uint40) {
+        return (address(uint160(_proposedPauseAuthorityChange)), uint40(_proposedPauseAuthorityChange >> 160));
+    }
+
+    /// @dev Proposes a new pause authority with a timelock delay.
+    function proposePauseAuthorityChange(address newPauseAuthority) public virtual {
```
Comment (legion2002) [inline src/PauseAuthority.sol:88]: Can we shorten function names, here proposeNewAuthority, getProposedAuthority, executeNewAuthority ?

File: src/PauseAuthority.sol (hunk 1 above)
Comment (NehharShah) [reply src/PauseAuthority.sol:123]: sure

File: src/PauseAuthority.sol (hunk 4 above)
Comment (NehharShah) [reply src/PauseAuthority.sol:88]: I will try my best :)

File: src/PauseAuthority.sol [hunk 5]
```diff
@@ -5,49 +5,67 @@ abstract contract PauseAuthority {
...
+            if (msg.sender != authority || block.timestamp < lastUnpaused + PAUSE_COOLDOWN || pauseFlag == 1) {
                 revert Unauthorized();
             }
 
             // Set the pause flag.
             pauseFlag = 1;
-            _pauseConfig = (block.timestamp << 160) | uint256(uint160(authority));
+            // Don't update lastUnpaused timestamp when pausing
         } else {
             if (msg.sender == authority || block.timestamp > timeout) {
```
Comment (legion2002) [inline src/PauseAuthority.sol:65]: wait, this is wrong I think.
This checks that the contracts can be unpaused after 4 weeks of the last unpause.
We want it to check that the contracts can be unpaused after 4 weeks of last pause. We need this to prevent censorship by the pause authority.

I think we'll have to set the 40 bits to lastPaused when pauseFlag = true, and lastUnpaused when pauseFlag = false.

**Review by legion2002 — CHANGES_REQUESTED** [review]

File: src/PauseAuthority.sol (hunk 5 above)
Comment (NehharShah) [reply src/PauseAuthority.sol:65]: Okay, let me correct this

**Comment by gakonst** [comment]
Id like such changes to the core mechanics to wait for my review @legion2002 pls don't merge
